BlueprintUICommonControls/Sources/Internal/KeyboardObserver.swift

```diff
@@ -0,0 +1,172 @@
+ 
+ iOS Docs for keyboard management:
+ https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/KeyboardManagement/KeyboardManagement.html
+ */
+final class KeyboardObserver
+{
+    private let center : NotificationCenter
+    
+    weak var delegate : KeyboardObserverDelegate?
+    
+    //
+    // MARK: Initialization
+    //
+    
+    init(center : NotificationCenter = .default)
+    {
+        self.center = center
+
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardWillChangeFrameNotification, object: nil)
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardDidChangeFrameNotification, object: nil)
+    }
+    
+    private var latestNotification : NotificationInfo?
+    
+    //
+    // MARK: Handling Changes
+    //
+    
+    enum KeyboardFrame : Equatable
+    {
+        case notVisible
+        case visible(frame: CGRect)
+    }
+    
+    func currentFrame(in view : UIView) -> KeyboardFrame?
+    {
+        guard view.window != nil else {
+            return nil
+        }
+        
+        guard let notification = self.latestNotification else {
+            return nil
+        }
+        
+        let frame = view.convert(notification.endingFrame, from: nil)
+        
+        if frame.intersects(view.bounds) {
+            return .visible(frame: frame)
+        } else {
+            return .notVisible
+        }
+    }
+    
+    private func animate(with info : NotificationInfo, _ animations : @escaping () -> ())
+    {
+        if info.animationDuration > 0.0 {
+            /**
+             Create an animation curve with the correct curve for showing or hiding the keyboard.
+             
+             This is unfortunately a private UIView curve. However, we can map it to the animation options' curve
+             like so: https://stackoverflow.com/questions/26939105/keyboard-animation-curve-as-int
+             */
+            let animationOptions = UIView.AnimationOptions(rawValue: info.animationCurve << 16)
+            
+            UIView.animate(
+                withDuration: info.animationDuration,
+                delay: 0.0,
+                options: animationOptions,
+                animations: animations
+            )
+        } else {
+            animations()
+        }
+    }
+    
+    //
+    // MARK: Receiving Updates
+    //
+    
+    private func receivedUpdatedKeyboardInfo(_ new : NotificationInfo)
+    {
+        let old = self.latestNotification
+        
+        self.latestNotification = new
+        
+        if let old = old {
+            guard old.endingFrame != new.endingFrame else {
+                return
+            }
+        }
+        
+        self.animate(with: new) {
+            self.delegate?.keyboardFrameWillChange(observer: self)
```

Comment:
It seems weird to me that this class wraps delegate calls in animation block. That's a lot of responsibility for an "observer". The decision of "how to react to this keyboard frame change" could probably be left up to the delegate.

Reply:
I agree that this is unexpected. I've always modeled the animation attributes in a type that is vended to the delegate (`keyboardFrameWillChange(to:animationAttriubutes:)` or something).

Reply:
One reason for this is because the animation curve for keyboards is a private value you need to get by shifting the curve by << 16. I'll fix by passing in the pre-shifted value for the delegate to use?

Reply:
Yeah, I've kept the conversion internal and provided the final value. `KeyboardObserver` isn't public though, so these API quirks don't leak out.